File: frontend/services/social-service.ts
/**
 * Social Service
 * 社交功能服务层 - 处理所有社交相关的 API 调用
 */

import { api } from '@/lib/api-client';
import type {
  FollowUser,
  FollowResponse,
  LikeResponse,
  BookmarkResponse,
  BookmarkItem,
  BookmarkFolder,
  Notification,
  NotificationStats,
  NotificationPreferences,
  UserSocialProfile,
  RelationshipStats,
  ActivityFeed,
  FeedResponse,
  SocialApiResponse,
} from '@/types/social.types';

const API_BASE = '/api/social';

// ============================================================================
// Follow Service - 关注功能
// ============================================================================

/**
 * 关注用户
 */
export async function followUser(userId: string): Promise<FollowResponse> {
  const response = await api.post<FollowResponse>(
    `${API_BASE}/follow`,
    { userId }
  );
  return response.data;
}

/**
 * 取消关注用户
 */
export async function unfollowUser(userId: string): Promise<FollowResponse> {
  const response = await api.post<FollowResponse>(
    `${API_BASE}/unfollow`,
    { userId }
  );
  return response.data;
}

/**
 * 批量关注用户
 */
export async function followMultipleUsers(userIds: string[]): Promise<FollowResponse[]> {
  const requests = userIds.map(userId =>
    () => api.post<FollowResponse>(`${API_BASE}/follow`, { userId })
  );
  return api.batch(requests);
}

/**
 * 获取用户的关注列表
 */
export async function getFollowing(
  userId: string,
  page = 1,
  perPage = 20
): Promise<{ users: FollowUser[]; total: number }> {
  const response = await api.get<{
    users: FollowUser[];
    total: number;
  }>(`${API_BASE}/following/${userId}`, {
    params: { page, per_page: perPage },
  });
  return response.data;
}

/**
 * 获取用户的粉丝列表
 */
export async function getFollowers(
  userId: string,
  page = 1,
  perPage = 20
): Promise<{ users: FollowUser[]; total: number }> {
  const response = await api.get<{
    users: FollowUser[];
    total: number;
  }>(`${API_BASE}/followers/${userId}`, {
    params: { page, per_page: perPage },
  });
  return response.data;
}

/**
 * 获取推荐关注的用户
 */
export async function getRecommendedUsers(limit = 10): Promise<FollowUser[]> {
  const response = await api.get<FollowUser[]>(
    `${API_BASE}/recommendations`,
    { params: { limit } }
  );
  return response.data;
}

// ============================================================================
// Like Service - 点赞功能
// ============================================================================

/**
 * 点赞文章或评论
 */
export async function likeContent(
  targetId: string,
  targetType: 'post' | 'comment'
): Promise<LikeResponse> {
  const response = await api.post<LikeResponse>(
    `${API_BASE}/like`,
    { targetId, targetType }
  );
  return response.data;
}

/**
 * 取消点赞
 */
export async function unlikeContent(
  targetId: string,
  targetType: 'post' | 'comment'
): Promise<LikeResponse> {
  const response = await api.post<LikeResponse>(
    `${API_BASE}/unlike`,
    { targetId, targetType }
  );
  return response.data;
}

/**
 * 检查是否已点赞
 */
export async function checkLikeStatus(
  targetId: string,
  targetType: 'post' | 'comment'
): Promise<{ isLiked: boolean }> {
  const response = await api.get<{ isLiked: boolean }>(
    `${API_BASE}/like/status`,
    { params: { targetId, targetType } }
  );
  return response.data;
}

/**
 * 获取点赞列表
 */
export async function getLikes(
  targetId: string,
  targetType: 'post' | 'comment',
  page = 1,
  perPage = 20
): Promise<{ users: Array<{ id: string; username: string; displayName: string; avatar?: string }>; total: number }> {
  const response = await api.get<{
    users: Array<{ id: string; username: string; displayName: string; avatar?: string }>;
    total: number;
  }>(`${API_BASE}/likes/${targetType}/${targetId}`, {
    params: { page, per_page: perPage },
  });
  return response.data;
}

// ============================================================================
// Bookmark Service - 书签功能
// ============================================================================

/**
 * 添加书签
 */
export async function addBookmark(
  targetId: string,
  targetType: 'post' | 'comment',
  folderId?: string
): Promise<BookmarkResponse> {
  const response = await api.post<BookmarkResponse>(
    `${API_BASE}/bookmark`,
    { targetId, targetType, folderId }
  );
  return response.data;
}

/**
 * 移除书签
 */
export async function removeBookmark(
  bookmarkId: string
): Promise<BookmarkResponse> {
  const response = await api.delete<BookmarkResponse>(
    `${API_BASE}/bookmark/${bookmarkId}`
  );
  return response.data;
}

/**
 * 获取书签列表
 */
export async function getBookmarks(
  page = 1,
  perPage = 20,
  folderId?: string
): Promise<{ bookmarks: BookmarkItem[]; total: number }> {
  const response = await api.get<{
    bookmarks: BookmarkItem[];
    total: number;
  }>(`${API_BASE}/bookmarks`, {
    params: { page, per_page: perPage, folderId },
  });
  return response.data;
}

/**
 * 创建书签文件夹
 */
export async function createBookmarkFolder(
  name: string,
  icon?: string,
  color?: string
): Promise<BookmarkFolder> {
  const response = await api.post<BookmarkFolder>(
    `${API_BASE}/bookmark/folder`,
    { name, icon, color }
  );
  return response.data;
}

/**
 * 更新书签文件夹
 */
export async function updateBookmarkFolder(
  folderId: string,
  updates: Partial<Pick<BookmarkFolder, 'name' | 'icon' | 'color'>>
): Promise<BookmarkFolder> {
  const response = await api.put<BookmarkFolder>(
    `${API_BASE}/bookmark/folder/${folderId}`,
    updates
  );
  return response.data;
}

/**
 * 删除书签文件夹
 */
export async function deleteBookmarkFolder(folderId: string): Promise<void> {
  await api.delete(`${API_BASE}/bookmark/folder/${folderId}`);
}

/**
 * 获取书签文件夹列表
 */
export async function getBookmarkFolders(): Promise<BookmarkFolder[]> {
  const response = await api.get<BookmarkFolder[]>(
    `${API_BASE}/bookmark/folders`
  );
  return response.data;
}

// ============================================================================
// Notification Service - 通知功能
// ============================================================================

/**
 * 获取通知列表
 */
export async function getNotifications(
  page = 1,
  perPage = 20,
  unreadOnly = false
): Promise<{ notifications: Notification[]; total: number }> {
  const response = await api.get<{
    notifications: Notification[];
    total: number;
  }>(`${API_BASE}/notifications`, {
    params: { page, per_page: perPage, unreadOnly },
  });
  return response.data;
}

/**
 * 获取通知统计
 */
export async function getNotificationStats(): Promise<NotificationStats> {
  const response = await api.get<NotificationStats>(
    `${API_BASE}/notifications/stats`
  );
  return response.data;
}

/**
 * 标记通知为已读
 */
export async function markNotificationAsRead(notificationId: string): Promise<void> {
  await api.put(`${API_BASE}/notifications/${notificationId}/read`);
}

/**
 * 批量标记通知为已读
 */
export async function markAllNotificationsAsRead(): Promise<void> {
  await api.put(`${API_BASE}/notifications/read-all`);
}

/**
 * 删除通知
 */
export async function deleteNotification(notificationId: string): Promise<void> {
  await api.delete(`${API_BASE}/notifications/${notificationId}`);
}

/**
 * 获取通知偏好设置
 */
export async function getNotificationPreferences(): Promise<NotificationPreferences> {
  const response = await api.get<NotificationPreferences>(
    `${API_BASE}/notifications/preferences`
  );
  return response.data;
}

/**
 * 更新通知偏好设置
 */
export async function updateNotificationPreferences(
  preferences: Partial<NotificationPreferences>
): Promise<NotificationPreferences> {
  const response = await api.put<NotificationPreferences>(
    `${API_BASE}/notifications/preferences`,
    preferences
  );
  return response.data;
}

// ============================================================================
// User Profile Service - 用户资料
// ============================================================================

/**
 * 获取用户社交资料
 */
export async function getUserSocialProfile(
  userId: string
): Promise<UserSocialProfile> {
  const response = await api.get<UserSocialProfile>(
    `${API_BASE}/profile/${userId}`
  );
  return response.data;
}

/**
 * 获取用户关系统计
 */
export async function getUserRelationshipStats(
  userId: string
): Promise<RelationshipStats> {
  const response = await api.get<RelationshipStats>(
    `${API_BASE}/profile/${userId}/stats`
  );
  return response.data;
}

/**
 * 更新用户资料
 */
export async function updateUserProfile(
  updates: Partial<Pick<UserSocialProfile, 'bio' | 'location' | 'website' | 'displayName'>>
): Promise<UserSocialProfile> {
  const response = await api.put<UserSocialProfile>(
    `${API_BASE}/profile`,
    updates
  );
  return response.data;
}

/**
 * 上传头像
 */
export async function uploadAvatar(file: File): Promise<{ avatarUrl: string }> {
  const response = await api.upload<{ avatarUrl: string }>(
    `${API_BASE}/profile/avatar`,
    file
  );
  return response.data;
}

// ============================================================================
// Activity & Feed Service - 动态和 Feed
// ============================================================================

/**
 * 获取用户动态
 */
export async function getUserActivity(
  userId?: string,
  page = 1,
  perPage = 20
): Promise<ActivityFeed> {
  const response = await api.get<ActivityFeed>(
    `${API_BASE}/activity`,
    { params: { userId, page, per_page: perPage } }
  );
  return response.data;
}

/**
 * 获取 Feed
 */
export async function getFeed(
  type: 'global' | 'following' | 'personal' = 'following',
  page = 1,
  perPage = 20
): Promise<FeedResponse> {
  const response = await api.get<FeedResponse>(
    `${API_BASE}/feed/${type}`,
    { params: { page, per_page: perPage } }
  );
  return response.data;
}

// ============================================================================
// Blocking & Muting - 屏蔽和静音
// ============================================================================

/**
 * 屏蔽用户
 */
export async function blockUser(userId: string): Promise<void> {
  await api.post(`${API_BASE}/block`, { userId });
}

/**
 * 取消屏蔽用户
 */
export async function unblockUser(userId: string): Promise<void> {
  await api.post(`${API_BASE}/unblock`, { userId });
}

/**
 * 静音用户
 */
export async function muteUser(userId: string): Promise<void> {
  await api.post(`${API_BASE}/mute`, { userId });
}

/**
 * 取消静音用户
 */
export async function unmuteUser(userId: string): Promise<void> {
  await api.post(`${API_BASE}/unmute`, { userId });
}

/**
 * 获取已屏蔽用户列表
 */
export async function getBlockedUsers(page = 1, perPage = 20): Promise<{
  users: UserSocialProfile[];
  total: number;
}> {
  const response = await api.get<{
    users: UserSocialProfile[];
    total: number;
  }>(`${API_BASE}/blocked`, {
    params: { page, per_page: perPage },
  });
  return response.data;
}

/**
 * 获取已静音用户列表
 */
export async function getMutedUsers(page = 1, perPage = 20): Promise<{
  users: UserSocialProfile[];
  total: number;
}> {
  const response = await api.get<{
    users: UserSocialProfile[];
    total: number;
  }>(`${API_BASE}/muted`, {
    params: { page, per_page: perPage },
  });
  return response.data;
}

File: backend/app/services/social_service.py
"""
Social Service
Business logic for social features
"""

from typing import List, Optional, Tuple
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, desc, func
from datetime import datetime, timedelta

from app.models.social import Follow, Like, Bookmark, BookmarkFolder, Notification, NotificationPreference, Activity
from app.models.user import User
from app.schemas.social import (
    FollowCreate, FollowStats, FollowUser, FollowListResponse,
    LikeCreate, LikeResponse, LikeStats,
    BookmarkCreate, BookmarkFolderCreate, BookmarkFolderUpdate,
    BookmarkResponse, BookmarkItem, BookmarksListResponse, BookmarkFolderResponse,
    NotificationCreate, NotificationListResponse, NotificationStats,
    NotificationPreferences, NotificationPreferencesUpdate,
    ActivityFeedResponse
)


class SocialService:
    """Social features service"""

    # ========================================================================
    # Follow Methods
    # ========================================================================

    @staticmethod
    def follow_user(db: Session, follower_id: int, following_id: int) -> Tuple[bool, str]:
        """Follow a user"""
        # Check if trying to follow self
        if follower_id == following_id:
            return False, "Cannot follow yourself"

        # Check if already following
        existing = db.query(Follow).filter(
            and_(
                Follow.follower_id == follower_id,
                Follow.following_id == following_id
            )
        ).first()

        if existing:
            return False, "Already following this user"

        # Create follow relationship
        follow = Follow(
            follower_id=follower_id,
            following_id=following_id
        )
        db.add(follow)

        # Create notification
        SocialService._create_notification(
            db,
            user_id=following_id,
            type="follow",
            title="New follower",
            message=f"User #${follower_id} started following you",
            actor_id=follower_id,
            target_id=following_id,
            target_type="user"
        )

        # Create activity
        SocialService._create_activity(
            db,
            user_id=follower_id,
            type="follow",
            actor_id=follower_id,
            target_id=following_id,
            target_type="user"
        )

        db.commit()
        return True, "Successfully followed user"

    @staticmethod
    def unfollow_user(db: Session, follower_id: int, following_id: int) -> Tuple[bool, str]:
        """Unfollow a user"""
        follow = db.query(Follow).filter(
            and_(
                Follow.follower_id == follower_id,
                Follow.following_id == following_id
            )
        ).first()

        if not follow:
            return False, "Not following this user"

        db.delete(follow)
        db.commit()
        return True, "Successfully unfollowed user"

    @staticmethod
    def get_follow_stats(db: Session, user_id: int, current_user_id: Optional[int] = None) -> FollowStats:
        """Get user follow statistics"""
        followers_count = db.query(func.count(Follow.id)).filter(
            Follow.following_id == user_id
        ).scalar() or 0

        following_count = db.query(func.count(Follow.id)).filter(
            Follow.follower_id == user_id
        ).scalar() or 0

        is_following = False
        if current_user_id and current_user_id != user_id:
            existing = db.query(Follow).filter(
                and_(
                    Follow.follower_id == current_user_id,
                    Follow.following_id == user_id
                )
            ).first()
            is_following = existing is not None

        return FollowStats(
            user_id=user_id,
            followers_count=followers_count,
            following_count=following_count,
            is_following=is_following
        )

    @staticmethod
    def get_followers(db: Session, user_id: int, page: int = 1, page_size: int = 20) -> FollowListResponse:
        """Get user's followers"""
        query = db.query(Follow).filter(Follow.following_id == user_id)
        total = query.count()

        offset = (page - 1) * page_size
        follows = query.order_by(desc(Follow.created_at)).offset(offset).limit(page_size).all()

        items = []
        for follow in follows:
            user = db.query(User).filter(User.id == follow.follower_id).first()
            if user:
                items.append(FollowUser(
                    id=user.id,
                    username=user.username,
                    display_name=user.display_name or user.username,
                    avatar=user.avatar_url,
                    bio=user.bio,
                    verified=False,
                    followers_count=0,  # Can be optimized with joins
                    is_following=False,
                    followed_at=follow.created_at
                ))

        return FollowListResponse(
            items=items,
            total=total,
            page=page,
            page_size=page_size,
            has_more=offset + page_size < total
        )

    @staticmethod
    def get_following(db: Session, user_id: int, page: int = 1, page_size: int = 20) -> FollowListResponse:
        """Get users that user is following"""
        query = db.query(Follow).filter(Follow.follower_id == user_id)
        total = query.count()

        offset = (page - 1) * page_size
        follows = query.order_by(desc(Follow.created_at)).offset(offset).limit(page_size).all()

        items = []
        for follow in follows:
            user = db.query(User).filter(User.id == follow.following_id).first()
            if user:
                items.append(FollowUser(
                    id=user.id,
                    username=user.username,
                    display_name=user.display_name or user.username,
                    avatar=user.avatar_url,
                    bio=user.bio,
                    verified=False,
                    followers_count=0,
                    is_following=True,
                    followed_at=follow.created_at
                ))

        return FollowListResponse(
            items=items,
            total=total,
            page=page,
            page_size=page_size,
            has_more=offset + page_size < total
        )

    # ========================================================================
    # Like Methods
    # ========================================================================

    @staticmethod
    def like_item(db: Session, user_id: int, target_id: int, target_type: str) -> LikeResponse:
        """Like a post or comment"""
        # Check if already liked
        existing = db.query(Like).filter(
            and_(
                Like.user_id == user_id,
                Like.target_id == target_id,
                Like.target_type == target_type
            )
        ).first()

        if existing:
            # Unlike (toggle off)
            db.delete(existing)
            db.commit()
            likes_count = db.query(func.count(Like.id)).filter(
                and_(
                    Like.target_id == target_id,
                    Like.target_type == target_type
                )
            ).scalar() or 0

            return LikeResponse(
                success=True,
                liked=False,
                likes_count=likes_count,
                message="Removed like"
            )

        # Create like
        like = Like(
            user_id=user_id,
            target_id=target_id,
            target_type=target_type
        )
        db.add(like)

        # Get likes count
        likes_count = db.query(func.count(Like.id)).filter(
            and_(
                Like.target_id == target_id,
                Like.target_type == target_type
            )
        ).scalar() or 1

        # Create notification for post owner
        if target_type == "post":
            # Get post owner
            post = db.query(Post).filter(Post.id == target_id).first()
            if post and post.author_id != user_id:
                SocialService._create_notification(
                    db,
                    user_id=post.author_id,
                    type="like",
                    title="Your post was liked",
                    message=f"Someone liked your post",
                    actor_id=user_id,
                    target_id=target_id,
                    target_type="post"
                )

        db.commit()
        return LikeResponse(
            success=True,
            liked=True,
            likes_count=likes_count,
            message="Added like"
        )

    @staticmethod
    def get_like_stats(db: Session, target_id: int, target_type: str, user_id: Optional[int] = None) -> LikeStats:
        """Get like statistics for a target"""
        likes_count = db.query(func.count(Like.id)).filter(
            and_(
                Like.target_id == target_id,
                Like.target_type == target_type
            )
        ).scalar() or 0

        is_liked = False
        if user_id:
            existing = db.query(Like).filter(
                and_(
                    Like.user_id == user_id,
                    Like.target_id == target_id,
                    Like.target_type == target_type
                )
            ).first()
            is_liked = existing is not None

        return LikeStats(
            likes_count=likes_count,
            is_liked=is_liked
        )

    # ========================================================================
    # Bookmark Methods
    # ========================================================================

    @staticmethod
    def create_bookmark_folder(db: Session, user_id: int, data: BookmarkFolderCreate) -> BookmarkFolder:
        """Create a bookmark folder"""
        folder = BookmarkFolder(
            user_id=user_id,
            name=data.name,
            icon=data.icon,
            color=data.color,
            description=data.description,
            is_default=False
        )
        db.add(folder)
        db.commit()
        db.refresh(folder)
        return folder

    @staticmethod
    def add_bookmark(db: Session, user_id: int, data: BookmarkCreate) -> BookmarkResponse:
        """Add a bookmark"""
        # Check if already bookmarked
        existing = db.query(Bookmark).filter(
            and_(
                Bookmark.user_id == user_id,
                Bookmark.target_id == data.target_id,
                Bookmark.target_type == data.target_type
            )
        ).first()

        if existing:
            # Remove bookmark (toggle off)
            db.delete(existing)
            db.commit()
            bookmarks_count = db.query(func.count(Bookmark.id)).filter(
                Bookmark.user_id == user_id
            ).scalar() or 0

            return BookmarkResponse(
                success=True,
                bookmarked=False,
                bookmarks_count=bookmarks_count,
                message="Removed bookmark"
            )

        # Create bookmark
        bookmark = Bookmark(
            user_id=user_id,
            target_id=data.target_id,
            target_type=data.target_type,
            folder_id=data.folder_id,
            notes=data.notes
        )
        db.add(bookmark)

        # Get bookmarks count
        bookmarks_count = db.query(func.count(Bookmark.id)).filter(
            Bookmark.user_id == user_id
        ).scalar() or 1

        db.commit()
        return BookmarkResponse(
            success=True,
            bookmarked=True,
            bookmarks_count=bookmarks_count,
            message="Added bookmark"
        )

    # ========================================================================
    # Notification Methods
    # ========================================================================

    @staticmethod
    def _create_notification(
        db: Session,
        user_id: int,
        type: str,
        title: str,
        message: str,
        actor_id: Optional[int] = None,
        target_id: Optional[int] = None,
        target_type: Optional[str] = None,
        data: Optional[dict] = None
    ) -> Notification:
        """Create a notification (internal)"""
        import json

        notification = Notification(
            user_id=user_id,
            type=type,
            title=title,
            message=message,
            actor_id=actor_id,
            target_id=target_id,
            target_type=target_type,
            data=json.dumps(data) if data else None
        )
        db.add(notification)
        return notification

    @staticmethod
    def get_notifications(
        db: Session,
        user_id: int,
        page: int = 1,
        page_size: int = 20,
        type_filter: Optional[str] = None,
        unread_only: bool = False
    ) -> NotificationListResponse:
        """Get user notifications"""
        query = db.query(Notification).filter(Notification.user_id == user_id)

        if type_filter:
            query = query.filter(Notification.type == type_filter)

        if unread_only:
            query = query.filter(Notification.is_read == False)

        # Don't show muted notifications
        query = query.filter(Notification.is_muted == False)

        total = query.count()
        unread_count = db.query(func.count(Notification.id)).filter(
            and_(
                Notification.user_id == user_id,
                Notification.is_read == False
            )
        ).scalar() or 0

        offset = (page - 1) * page_size
        notifications = query.order_by(desc(Notification.created_at)).offset(offset).limit(page_size).all()

        items = []
        for notification in notifications:
            import json
            data = None
            if notification.data:
                try:
                    data = json.loads(notification.data)
                except:
                    pass

            items.append(Notification(
                id=notification.id,
                user_id=notification.user_id,
                type=notification.type,
                title=notification.title,
                message=notification.message,
                data=data,
                actor_id=notification.actor_id,
                target_id=notification.target_id,
                target_type=notification.target_type,
                is_read=notification.is_read,
                is_muted=notification.is_muted,
                created_at=notification.created_at,
                updated_at=notification.updated_at
            ))

        return NotificationListResponse(
            items=items,
            total=total,
            unread=unread_count,
            page=page,
            page_size=page_size,
            has_more=offset + page_size < total
        )

    @staticmethod
    def mark_notification_read(db: Session, notification_id: int, user_id: int) -> bool:
        """Mark notification as read"""
        notification = db.query(Notification).filter(
            and_(
                Notification.id == notification_id,
                Notification.user_id == user_id
            )
        ).first()

        if not notification:
            return False

        notification.is_read = True
        notification.updated_at = datetime.utcnow()
        db.commit()
        return True

    @staticmethod
    def mark_all_read(db: Session, user_id: int) -> int:
        """Mark all notifications as read"""
        count = db.query(Notification).filter(
            and_(
                Notification.user_id == user_id,
                Notification.is_read == False
            )
        ).update({
            "is_read": True,
            "updated_at": datetime.utcnow()
        })
        db.commit()
        return count

    @staticmethod
    def get_notification_stats(db: Session, user_id: int) -> NotificationStats:
        """Get notification statistics"""
        total = db.query(func.count(Notification.id)).filter(
            Notification.user_id == user_id
        ).scalar() or 0

        unread = db.query(func.count(Notification.id)).filter(
            and_(
                Notification.user_id == user_id,
                Notification.is_read == False
            )
        ).scalar() or 0

        return NotificationStats(
            total=total,
            unread=unread,
            read=total - unread
        )

    # ========================================================================
    # Activity Methods
    # ========================================================================

    @staticmethod
    def _create_activity(
        db: Session,
        user_id: int,
        type: str,
        actor_id: Optional[int] = None,
        target_id: Optional[int] = None,
        target_type: Optional[str] = None,
        metadata: Optional[dict] = None
    ) -> Activity:
        """Create an activity (internal)"""
        import json

        activity = Activity(
            user_id=user_id,
            type=type,
            actor_id=actor_id,
            target_id=target_id,
            target_type=target_type,
            metadata=json.dumps(metadata) if metadata else None
        )
        db.add(activity)
        return activity

    @staticmethod
    def get_activity_feed(
        db: Session,
        user_id: int,
        page: int = 1,
        page_size: int = 20
    ) -> ActivityFeedResponse:
        """Get user activity feed"""
        query = db.query(Activity).filter(Activity.user_id == user_id)
        total = query.count()

        offset = (page - 1) * page_size
        activities = query.order_by(desc(Activity.created_at)).offset(offset).limit(page_size).all()

        items = []
        for activity in activities:
            import json
            metadata = None
            if activity.metadata:
                try:
                    metadata = json.loads(activity.metadata)
                except:
                    pass

            items.append(Activity(
                id=activity.id,
                user_id=activity.user_id,
                type=activity.type,
                actor_id=activity.actor_id,
                target_id=activity.target_id,
                target_type=activity.target_type,
                metadata=metadata,
                created_at=activity.created_at
            ))

        return ActivityFeedResponse(
            items=items,
            total=total,
            page=page,
            page_size=page_size,
            has_more=offset + page_size < total
        )
